File: go/linked_list/addTwoNumbers2.go
package main

import (
	. "wtf-leetcode/linked_list/def"
)

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	p1, p2 := l1, l2
	preNode := &ListNode{}
	p := preNode
	carry := 0
	for p1 != nil && p2 != nil {
		curr := (p1.Val + p2.Val + carry) % 10
		carry = (p1.Val + p2.Val + carry) / 10
		p.Next = &ListNode{Val: curr}
		p = p.Next
		p1 = p1.Next
		p2 = p2.Next
	}
	for p1 != nil {
		curr := (p1.Val + carry) % 10
		carry = (p1.Val + carry) / 10
		p.Next = &ListNode{Val: curr}
		p = p.Next
		p1 = p1.Next
	}
	for p2 != nil {
		curr := (p2.Val + carry) % 10
		carry = (p2.Val + carry) / 10
		p.Next = &ListNode{Val: curr}
		p = p.Next
		p2 = p2.Next
	}
	if carry > 0 {
		p.Next = &ListNode{Val: carry}
	}
	return preNode.Next
}

func main() {
	l1 := &ListNode{Val: 2, Next: &ListNode{Val: 4, Next: &ListNode{Val: 3}}}
	l2 := &ListNode{Val: 5, Next: &ListNode{Val: 6, Next: &ListNode{Val: 4}}}
	PrintList(addTwoNumbers(l1, l2))
}

File: python/listed_list/addTwoNumbers2.py
from lib.node import *

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        dummy = l3 = ListNode(-1)
        while l1 and l2:
            val = l1.val + l2.val + carry
            carry = val // 10
            val = val % 10
            l3.next = ListNode(val)
            l3 = l3.next
            l1 = l1.next
            l2 = l2.next

        while l1:
            val = l1.val + carry
            carry = val // 10
            val = val % 10
            l3.next = ListNode(val)
            l3 = l3.next
            l1 = l1.next

        while l2:
            val = l2.val + carry
            carry = val // 10
            val = val % 10
            l3.next = ListNode(val)
            l3 = l3.next
            l2 = l2.next

        if carry:
            l3.next = ListNode(carry)

        return dummy.next

    def addTwoNumbers2(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        dummy = l3 = ListNode(-1)
        while l1 or l2 or carry:
            val = 0
            if l1:
                val += l1.val
                l1 = l1.next
            if l2:
                val += l2.val
                l2 = l2.next

            val += carry
            carry = val // 10
            val = val % 10
            l3.next = ListNode(val)
            l3 = l3.next

        return dummy.next

if __name__ == '__main__':
    s = Solution()
    # l1 = ListNode(2, ListNode(4, ListNode(3)))
    # l2 = ListNode(5, ListNode(6, ListNode(4)))

    l1 = ListNode(9, ListNode(9))
    l2 = ListNode(9, ListNode(9, ListNode(9)))
    print_nodes(s.addTwoNumbers2(l1, l2))